src/lib.rs

```diff
@@ -257,7 +257,10 @@ from_hex_array_impl! {
 #[must_use]
 #[cfg(feature = "alloc")]
 pub fn encode<T: AsRef<[u8]>>(data: T) -> String {
-    data.encode_hex()
+    let data = data.as_ref();
+    let mut out = vec![0; data.len() * 2];
+    encode_to_slice(data, &mut out).unwrap();
+    String::from_utf8(out).unwrap()
```

Comment:
When using from_utf8_unckecked here, it can improve performance by about 8%. (It is safe because we emit only hex characters.)
However, I didn't apply that change because I don't know the policy regarding the unsafe code in this crate.
If It is okay with using unsafe code, I'll add that change.

```
hex_encode              time:   [9.4555 us 9.4828 us 9.5106 us]                        
                        change: [-9.2447% -8.3200% -7.4582%] (p = 0.00 < 0.05)
                        Performance has improved.
```

Reply:
Some overlap with https://github.com/KokaKiwi/rust-hex/pull/66 here I think.  In particular I suspect the combination of `ExactSizeIterator` will allow us to collect into a `String` efficiently without indirecting through `Vec`.